From 090ff3caf0d7fc4e5afbe2adff988735c2c86d7b Mon Sep 17 00:00:00 2001
From: ChromeOS Toolchain Worker
 <crostc-worker@crostc-chrotomation.iam.gserviceaccount.com>
Date: Fri, 13 Sep 2024 16:47:06 -0700
Subject: [PATCH] Make _LIBCPP_MAKE_OVERRIDABLE_FUNCTION a no-op on baremetal

This is a variation on the previous patch set (reverted in crrev.com/c/7648042), which overrode
_LIBCPP_MAKE_OVERRIDABLE_FUNCTION_DETECTABLE

Baremetal targets tend to use custom linker scripts, which may not
include the __lcxx_override section.

patch.cherry: false
patch.metadata.author: Jordan Abrahams-Whitehead <ajordanr@google.com>
patch.platforms: chromiumos
patch.version_range.from: 584947
patch.version_range.until: null

BUG=b:363082822
BUG=b:485361709
TEST=sudo emerge cross-arm-none-eabi/libcxx
TEST=sudo emerge cross-armv7m-cros-eabi/libcxx

---
 libcxx/src/include/overridable_function.h | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/libcxx/src/include/overridable_function.h b/libcxx/src/include/overridable_function.h
index 0b43f271486c..fdcf96cde629 100644
--- a/libcxx/src/include/overridable_function.h
+++ b/libcxx/src/include/overridable_function.h
@@ -63,8 +63,12 @@
 // want to be defining special sections inside user's executables which use our headers.
 //
 
-#if defined(_LIBCPP_OBJECT_FORMAT_MACHO)
-
+// Baremetal targets tend to use custom linker scripts, which may not include
+// the __lcxx_override section.
+#if defined(LIBCXXABI_BAREMETAL)
+#  define _LIBCPP_CAN_DETECT_OVERRIDDEN_FUNCTION 0
+#  define _LIBCPP_OVERRIDABLE_FUNCTION(type, name, arglist) _LIBCPP_WEAK type name arglist
+#elif defined(_LIBCPP_OBJECT_FORMAT_MACHO)
 #  define _LIBCPP_CAN_DETECT_OVERRIDDEN_FUNCTION 1
 #  define _LIBCPP_OVERRIDABLE_FUNCTION(type, name, arglist)                                                            \
     __attribute__((__section__("__TEXT,__lcxx_override,regular,pure_instructions"))) _LIBCPP_WEAK type name arglist
-- 
2.39.5